File: src/importexport.py
# importexport.py
#
# Copyright 2026 Unknown
#
# This program is free software: you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation, either version 3 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program.  If not, see <https://www.gnu.org/licenses/>.
#
# SPDX-License-Identifier: GPL-3.0-or-later

import os
import sqlite3
import tempfile
import uuid
import zipfile
from datetime import datetime

from gi.repository import Adw, Gtk, GLib, Gio


class ImportExportMixin:
    """Import and export: MyExpenses backup, raw DB backup/restore."""

    def on_import_database(self, action, param):
        """Show file chooser for importing MyExpenses backup (zip or raw DB)"""
        dialog = Gtk.FileDialog()
        dialog.set_title("Import from MyExpenses")
        dialog.set_modal(True)

        # Set initial folder to home directory
        home_dir = GLib.get_home_dir()
        try:
            initial_folder = Gio.File.new_for_path(home_dir)
            dialog.set_initial_folder(initial_folder)
        except Exception:
            pass

        filters = Gio.ListStore.new(Gtk.FileFilter)

        backup_filter = Gtk.FileFilter()
        backup_filter.set_name("MyExpenses Backups")
        backup_filter.add_pattern("*.zip")
        backup_filter.add_pattern("BACKUP")
        backup_filter.add_pattern("*.db")
        backup_filter.add_pattern("*.sqlite")
        filters.append(backup_filter)

        all_filter = Gtk.FileFilter()
        all_filter.set_name("All Files")
        all_filter.add_pattern("*")
        filters.append(all_filter)

        dialog.set_filters(filters)
        dialog.open(self, None, self.on_import_file_selected)

    def on_import_file_selected(self, dialog, result):
        """Handle file selection for import"""
        try:
            file = dialog.open_finish(result)
            path = file.get_path()

            # If it's a zip, extract the BACKUP database to a temp file
            if zipfile.is_zipfile(path):
                path = self._extract_backup_from_zip(path)

            accounts_imported, transactions_imported = self.import_sqlite_database(path)

            success_dialog = Adw.MessageDialog.new(self)
            success_dialog.set_heading("Import Successful")
            success_dialog.set_body(f"Imported {accounts_imported} account(s) and {transactions_imported} transaction(s)")
            success_dialog.add_response("ok", "OK")
            success_dialog.set_response_appearance("ok", Adw.ResponseAppearance.DEFAULT)
            success_dialog.connect('response', lambda d, r: self.on_import_complete())
            success_dialog.present()

        except GLib.Error as e:
            error_dialog = Adw.MessageDialog.new(self)
            error_dialog.set_heading("Import Failed")
            error_dialog.set_body(f"Could not open file: {str(e)}")
            error_dialog.add_response("ok", "OK")
            error_dialog.present()
        except Exception as e:
            error_dialog = Adw.MessageDialog.new(self)
            error_dialog.set_heading("Import Failed")
            error_dialog.set_body(str(e))
            error_dialog.add_response("ok", "OK")
            error_dialog.present()

    def _extract_backup_from_zip(self, zip_path):
        """Extract the BACKUP database from a MyExpenses zip archive.

        Returns the path to the extracted file.
        """
        with zipfile.ZipFile(zip_path, 'r') as zf:
            names = zf.namelist()
            # Look for the BACKUP file (the actual SQLite database)
            backup_name = None
            for name in names:
                if name == 'BACKUP' or name.endswith('/BACKUP'):
                    backup_name = name
                    break

            if backup_name is None:
                raise ValueError(
                    "No BACKUP database found in the zip file.\n"
                    f"Contents: {', '.join(names)}"
                )

            tmp_dir = tempfile.mkdtemp(prefix='expenses-import-')
            zf.extract(backup_name, tmp_dir)
            extracted = os.path.join(tmp_dir, backup_name)

            # Guard against path traversal in malicious zip entries
            if not os.path.abspath(extracted).startswith(
                os.path.abspath(tmp_dir) + os.sep
            ):
                raise ValueError("Zip contains unsafe path")

            return extracted

    def import_sqlite_database(self, db_path):
        """Import data from SQLite database"""
        conn = None
        try:
            conn = sqlite3.connect(db_path)
            cursor = conn.cursor()

            # Fetch all accounts with opening balance
            cursor.execute("""
                SELECT _id, label, opening_balance FROM accounts
                ORDER BY sort_key ASC
            """)
            accounts = {}
            account_list = []
            opening_balances = {}

            for account_id, account_label, opening_balance_cents in cursor.fetchall():
                account_name = account_label.strip()
                account_list.append(account_name)
                accounts[account_id] = account_name
                opening_balances[account_id] = opening_balance_cents if opening_balance_cents else 0

            # Start with imported accounts or default if none
            if not account_list:
                account_list = ['Default']

            # Fetch all transactions with payee information
            cursor.execute("""
                SELECT t.account_id, t.amount, t.date, t.comment, p.name
                FROM transactions t
                LEFT JOIN payee p ON t.payee_id = p._id
                WHERE t.parent_id IS NULL
                ORDER BY t.date ASC
            """)

            expenses_data = {account: [] for account in account_list}

            # Add opening balance as initial transaction for each account
            for account_id, account_name in accounts.items():
                opening_balance_cents = opening_balances[account_id]
                if opening_balance_cents > 0:
                    opening_balance_euros = opening_balance_cents / 100.0
                    opening_expense = {
                        'amount': opening_balance_euros,
                        'payee': 'Opening Balance',
                        'note': 'Initial balance imported from database',
                        'date': '0000-01-01 00:00',
                        'is_opening_balance': True,
                        'is_income': True
                    }
                    expenses_data[account_name].insert(0, opening_expense)

            for account_id, amount_cents, timestamp, comment, payee_name in cursor.fetchall():
                # Get account name
                account_name = accounts.get(account_id, 'Default')

                # Convert amount from cents to euros
                amount_euros = amount_cents / 100.0
                is_income = amount_euros > 0

                # Convert timestamp (unix epoch in seconds) to ISO format
                try:
                    date_obj = datetime.fromtimestamp(timestamp)
                    date_str = date_obj.strftime('%Y-%m-%d %H:%M')
                except (ValueError, OSError, OverflowError):
                    date_str = datetime.now().strftime('%Y-%m-%d %H:%M')

                # Use payee name if available, otherwise fall back to default
                if payee_name:
                    payee = payee_name.strip()
                else:
                    payee = 'Unknown'

                expense = {
                    'amount': amount_euros,
                    'payee': payee,
                    'note': comment.strip() if comment else '',
                    'date': date_str,
                    'is_income': is_income
                }

                if account_name not in expenses_data:
                    expenses_data[account_name] = []

                expenses_data[account_name].append(expense)

            conn.close()

            # Replace all data via the database
            imported_data = {
                'accounts': account_list,
                'current_account': account_list[0] if account_list else 'Default',
                'expenses': expenses_data
            }
            self.db.replace_all_data(imported_data)
            self.current_account = imported_data['current_account']

            # Return counts for the success message
            total_transactions = sum(len(expenses) for expenses in expenses_data.values())
            return len(account_list), total_transactions

        except Exception as e:
            print(f"Error importing database: {e}")
            raise
        finally:
            if conn:
                conn.close()

    def on_import_complete(self):
        """Handle import completion - refresh UI"""
        self.setup_account_dropdown()
        self.update_expense_list()
        self.update_total()
        self.update_payee_suggestions()

    # ── MyExpenses-compatible export ───────────────────────────────────

    def on_export_myexpenses(self, action, param):
        """Export data as a MyExpenses-compatible backup zip"""
        dialog = Gtk.FileDialog()
        dialog.set_title("Export for MyExpenses")
        dialog.set_modal(True)

        now = datetime.now()
        dialog.set_initial_name(
            f"myexpenses-backup-{now.strftime('%Y%m%d-%H%M%S')}.zip"
        )

        filters = Gio.ListStore.new(Gtk.FileFilter)
        zip_filter = Gtk.FileFilter()
        zip_filter.set_name("Zip Archives")
        zip_filter.add_pattern("*.zip")
        filters.append(zip_filter)
        dialog.set_filters(filters)

        dialog.save(self, None, self._on_export_myexpenses_selected)

    def _on_export_myexpenses_selected(self, dialog, result):
        try:
            file = dialog.save_finish(result)
            dest_path = file.get_path()
        except GLib.Error:
            return  # user cancelled

        try:
            self._build_myexpenses_backup(dest_path)

            success = Adw.MessageDialog.new(self)
            success.set_heading("Export Successful")
            success.set_body(
                f"MyExpenses-compatible backup exported to:\n{dest_path}"
            )
            success.add_response("ok", "OK")
            success.present()

        except Exception as e:
            error = Adw.MessageDialog.new(self)
            error.set_heading("Export Failed")
            error.set_body(str(e))
            error.add_response("ok", "OK")
            error.present()

    def _build_myexpenses_backup(self, zip_path):
        """Build a MyExpenses-compatible BACKUP database and package as zip"""
        # Load schema from bundled gresource
        schema_bytes = Gio.resources_lookup_data(
            '/io/github/nico359/expenses/myexpenses_schema.sql',
            Gio.ResourceLookupFlags.NONE,
        )
        schema_sql = schema_bytes.get_data().decode('utf-8')

        with tempfile.TemporaryDirectory(prefix='expenses-export-') as tmp_dir:
            db_path = os.path.join(tmp_dir, 'BACKUP')
            conn = sqlite3.connect(db_path)
            conn.executescript(schema_sql)

            accounts = self.db.get_accounts()
            account_id_map = {}  # our account name → MyExpenses _id

            for name in accounts:
                our_id = self.db._get_account_id(name)

                # Get opening balance (stored as a pseudo-expense)
                row = self.db.conn.execute(
                    "SELECT amount, is_income FROM expenses "
                    "WHERE account_id = ? AND is_opening_balance = 1",
                    (our_id,),
                ).fetchone()
                opening_cents = round(row[0] * 100) if row else 0

                conn.execute(
                    "INSERT INTO accounts "
                    "(label, opening_balance, description, currency, type, "
                    " color, uuid) "
                    "VALUES (?, ?, '', 'EUR', 1, -3355444, ?)",
                    (name, opening_cents, str(uuid.uuid4())),
                )
                me_id = conn.execute(
                    "SELECT last_insert_rowid()"
                ).fetchone()[0]
                account_id_map[name] = me_id

            # Build payee lookup
            all_payees = self.db.get_all_payees()
            payee_id_map = {}
            for payee_name in all_payees:
                conn.execute(
                    "INSERT INTO payee (name, name_normalized) VALUES (?, ?)",
                    (payee_name, payee_name.lower()),
                )
                pid = conn.execute(
                    "SELECT last_insert_rowid()"
                ).fetchone()[0]
                payee_id_map[payee_name] = pid

            # Export transactions
            for name in accounts:
                our_id = self.db._get_account_id(name)
                me_account_id = account_id_map[name]

                rows = self.db.conn.execute(
                    "SELECT amount, payee, note, date, is_income "
                    "FROM expenses "
                    "WHERE account_id = ? AND is_opening_balance = 0 "
                    "ORDER BY id",
                    (our_id,),
                ).fetchall()

                for amount, payee, note, date_str, is_income in rows:
                    amount_cents = round(amount * 100)
                    if not is_income:
                        amount_cents = -amount_cents

                    try:
                        dt = datetime.strptime(date_str, '%Y-%m-%d %H:%M')
                    except ValueError:
                        dt = datetime.strptime(date_str, '%Y-%m-%d')
                    timestamp = int(dt.timestamp())

                    payee_id = payee_id_map.get(payee)

                    conn.execute(
                        "INSERT INTO transactions "
                        "(comment, date, value_date, amount, account_id, "
                        " payee_id, cr_status, number, uuid) "
                        "VALUES (?, ?, ?, ?, ?, ?, 'UNRECONCILED', '', ?)",
                        (note or '', timestamp, timestamp, amount_cents,
                         me_account_id, payee_id, str(uuid.uuid4())),
                    )

            conn.commit()
            conn.close()

            with zipfile.ZipFile(zip_path, 'w', zipfile.ZIP_DEFLATED) as zf:
                zf.write(db_path, 'BACKUP')

    # ── Database export / import ───────────────────────────────────────

    def on_export_database(self, action, param):
        """Export the database as a .db file"""
        dialog = Gtk.FileDialog()
        dialog.set_title("Export Database")
        dialog.set_modal(True)

        now = datetime.now()
        dialog.set_initial_name(
            f"expenses-backup-{now.strftime('%Y%m%d-%H%M%S')}.db"
        )

        filters = Gio.ListStore.new(Gtk.FileFilter)
        db_filter = Gtk.FileFilter()
        db_filter.set_name("SQLite Databases")
        db_filter.add_pattern("*.db")
        filters.append(db_filter)
        dialog.set_filters(filters)

        dialog.save(self, None, self._on_export_db_selected)

    def _on_export_db_selected(self, dialog, result):
        try:
            file = dialog.save_finish(result)
            dest_path = file.get_path()

            # Build the export in a temp file (writable), switch away from
            # WAL mode there, then copy the clean result to the portal path.
            tmp_fd, tmp_path = tempfile.mkstemp(suffix='.db')
            os.close(tmp_fd)
            try:
                dst = sqlite3.connect(tmp_path)
                self.db.conn.backup(dst)
                dst.execute("PRAGMA journal_mode=DELETE")
                dst.close()

                import shutil
                shutil.copy2(tmp_path, dest_path)
            finally:
                os.unlink(tmp_path)

            success = Adw.MessageDialog.new(self)
            success.set_heading("Export Successful")
            success.set_body(f"Database exported to:\n{dest_path}")
            success.add_response("ok", "OK")
            success.present()

        except GLib.Error:
            pass  # user cancelled
        except Exception as e:
            error = Adw.MessageDialog.new(self)
            error.set_heading("Export Failed")
            error.set_body(str(e))
            error.add_response("ok", "OK")
            error.present()

    def on_import_backup(self, action, param):
        """Import a previously exported .db file"""
        dialog = Gtk.FileDialog()
        dialog.set_title("Import Database")
        dialog.set_modal(True)

        filters = Gio.ListStore.new(Gtk.FileFilter)
        db_filter = Gtk.FileFilter()
        db_filter.set_name("SQLite Databases")
        db_filter.add_pattern("*.db")
        db_filter.add_pattern("*.sqlite")
        filters.append(db_filter)

        all_filter = Gtk.FileFilter()
        all_filter.set_name("All Files")
        all_filter.add_pattern("*")
        filters.append(all_filter)

        dialog.set_filters(filters)
        dialog.open(self, None, self._on_import_backup_selected)

    def _on_import_backup_selected(self, dialog, result):
        try:
            file = dialog.open_finish(result)
            path = file.get_path()
        except GLib.Error:
            return  # user cancelled

        # Ask for confirmation before replacing data
        confirm = Adw.MessageDialog.new(self)
        confirm.set_heading("Replace Existing Data?")
        confirm.set_body(
            "This will replace all current accounts, expenses and "
            "recurring definitions with the contents of the backup."
        )
        confirm.add_response("cancel", "Cancel")
        confirm.add_response("import", "Import")
        confirm.set_response_appearance(
            "import", Adw.ResponseAppearance.DESTRUCTIVE
        )
        confirm.set_default_response("cancel")
        confirm.connect("response", self._on_import_backup_confirmed, path)
        confirm.present()

    def _on_import_backup_confirmed(self, dialog, response, path):
        if response != "import":
            return

        try:
            # Open the backup file as immutable so SQLite won't try to
            # create -wal/-shm files next to it (the path may live on a
            # read-only Flatpak portal mount).
            from urllib.parse import quote
            uri = f"file:{quote(path, safe='/')}?immutable=1"
            src = sqlite3.connect(uri, uri=True)
            src.backup(self.db.conn)
            src.close()

            # Re-read settings
            accounts = self.db.get_accounts()
            self.current_account = self.db.get_setting(
                'current_account',
                accounts[0] if accounts else 'Default',
            )

            self.on_import_complete()

            success = Adw.MessageDialog.new(self)
            success.set_heading("Import Successful")
            success.set_body("Database restored from backup.")
            success.add_response("ok", "OK")
            success.present()

        except Exception as e:
            error = Adw.MessageDialog.new(self)
            error.set_heading("Import Failed")
            error.set_body(str(e))
            error.add_response("ok", "OK")
            error.present()

File: src/import_export.rs
/* import_export.rs
 *
 * Copyright 2026 nico359
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <https://www.gnu.org/licenses/>.
 *
 * SPDX-License-Identifier: GPL-3.0-or-later
 */

use gtk::prelude::*;
use adw::prelude::*;
use gtk::{gio, glib};
use rusqlite::{params, Connection};

use std::io::{Read, Write};
use std::path::Path;

use crate::db::Database;
use crate::window::ExpensesWindow;

impl ExpensesWindow {
    // --- Import from MyExpenses ---

    pub(crate) fn on_import_myexpenses(&self) {
        let dialog = gtk::FileDialog::builder()
            .title("Import from MyExpenses")
            .build();

        let filter = gtk::FileFilter::new();
        filter.add_pattern("*.zip");
        filter.add_pattern("BACKUP");
        filter.set_name(Some("MyExpenses backup (ZIP or BACKUP)"));
        let filters = gio::ListStore::new::<gtk::FileFilter>();
        filters.append(&filter);
        dialog.set_filters(Some(&filters));

        dialog.open(Some(self), None::<&gio::Cancellable>, glib::clone!(
            #[weak(rename_to = window)]
            self,
            move |result| {
                if let Ok(file) = result {
                    if let Some(path) = file.path() {
                        window.do_import_myexpenses(&path);
                    }
                }
            }
        ));
    }

    fn do_import_myexpenses(&self, path: &Path) {
        let path_str = path.to_string_lossy();

        let backup_path = if path_str.ends_with(".zip") {
            match Self::extract_backup_from_zip(path) {
                Ok(p) => p,
                Err(e) => {
                    self.show_error("Import Failed", &format!("Could not extract ZIP: {}", e));
                    return;
                }
            }
        } else {
            path.to_path_buf()
        };

        match self.import_myexpenses_db(&backup_path) {
            Ok((accounts, expenses)) => {
                self.db().replace_all_data(&accounts, &expenses).ok();
                self.setup_accounts();
                self.refresh_expenses();
                self.show_info("Import Complete", "MyExpenses data imported successfully.");
            }
            Err(e) => {
                self.show_error("Import Failed", &format!("Could not read database: {}", e));
            }
        }

        // Clean up temp file if we extracted from zip
        if path_str.ends_with(".zip") {
            std::fs::remove_file(&backup_path).ok();
        }
    }

    fn extract_backup_from_zip(zip_path: &Path) -> Result<std::path::PathBuf, String> {
        let file = std::fs::File::open(zip_path).map_err(|e| e.to_string())?;
        let mut archive = zip::ZipArchive::new(file).map_err(|e| e.to_string())?;

        for i in 0..archive.len() {
            let mut entry = archive.by_index(i).map_err(|e| e.to_string())?;
            let name = entry.name().to_string();
            if name == "BACKUP" || name.ends_with("/BACKUP") {
                let tmp_dir = std::env::temp_dir();
                let tmp_path = tmp_dir.join("expenses-import-BACKUP");
                let mut outfile = std::fs::File::create(&tmp_path).map_err(|e| e.to_string())?;
                let mut buf = Vec::new();
                entry.read_to_end(&mut buf).map_err(|e| e.to_string())?;
                outfile.write_all(&buf).map_err(|e| e.to_string())?;
                return Ok(tmp_path);
            }
        }
        Err("No BACKUP file found in ZIP archive".to_string())
    }

    fn import_myexpenses_db(
        &self,
        db_path: &Path,
    ) -> Result<(Vec<(String, i64)>, Vec<(i64, f64, String, String, String, bool, bool)>), String> {
        let conn = Connection::open(db_path).map_err(|e| e.to_string())?;

        // Read accounts
        let mut stmt = conn.prepare(
            "SELECT _id, label, opening_balance FROM accounts ORDER BY sort_key ASC"
        ).map_err(|e| e.to_string())?;

        let accounts_raw: Vec<(i64, String, i64)> = stmt.query_map([], |row| {
            Ok((row.get(0)?, row.get(1)?, row.get(2)?))
        }).map_err(|e| e.to_string())?
          .filter_map(|r| r.ok())
          .collect();

        let mut accounts = Vec::new();
        let mut account_id_map = std::collections::HashMap::new();
        let mut expenses = Vec::new();

        for (i, (old_id, name, opening_balance_cents)) in accounts_raw.iter().enumerate() {
            let new_id = (i + 1) as i64;
            account_id_map.insert(*old_id, new_id);
            accounts.push((name.clone(), i as i64));

            // Add opening balance as special transaction
            if *opening_balance_cents > 0 {
                let amount = *opening_balance_cents as f64 / 100.0;
                expenses.push((
                    new_id,
                    amount,
                    "Opening Balance".to_string(),
                    "Initial balance imported from database".to_string(),
                    "0000-01-01 00:00".to_string(),
                    true,  // is_income
                    true,  // is_opening_balance
                ));
            }
        }

        // Read transactions
        let mut stmt = conn.prepare(
            "SELECT t.account_id, t.amount, t.date, t.comment, p.name \
             FROM transactions t \
             LEFT JOIN payee p ON t.payee_id = p._id \
             WHERE t.parent_id IS NULL \
             ORDER BY t.date ASC"
        ).map_err(|e| e.to_string())?;

        let transactions: Vec<(i64, i64, i64, Option<String>, Option<String>)> = stmt.query_map([], |row| {
            Ok((
                row.get(0)?,
                row.get(1)?,
                row.get(2)?,
                row.get(3)?,
                row.get(4)?,
            ))
        }).map_err(|e| e.to_string())?
          .filter_map(|r| r.ok())
          .collect();

        for (old_account_id, amount_cents, timestamp, comment, payee_name) in &transactions {
            let new_account_id = match account_id_map.get(old_account_id) {
                Some(id) => *id,
                None => continue,
            };

            let amount_euros = (*amount_cents as f64 / 100.0).abs();
            let is_income = *amount_cents > 0;

            let date_str = match chrono::DateTime::from_timestamp(*timestamp, 0) {
                Some(dt) => dt.with_timezone(&chrono::Local).format("%Y-%m-%d %H:%M").to_string(),
                None => chrono::Local::now().format("%Y-%m-%d %H:%M").to_string(),
            };

            let payee = payee_name.as_deref().unwrap_or("").trim().to_string();
            let note = comment.as_deref().unwrap_or("").to_string();

            expenses.push((
                new_account_id,
                amount_euros,
                payee,
                note,
                date_str,
                is_income,
                false, // is_opening_balance
            ));
        }

        Ok((accounts, expenses))
    }

    // --- Export for MyExpenses ---

    pub(crate) fn on_export_myexpenses(&self) {
        let timestamp = chrono::Local::now().format("%Y-%m-%d_%H-%M-%S");
        let dialog = gtk::FileDialog::builder()
            .title("Export for MyExpenses")
            .initial_name(format!("myexpenses-export-{}.zip", timestamp))
            .build();

        dialog.save(Some(self), None::<&gio::Cancellable>, glib::clone!(
            #[weak(rename_to = window)]
            self,
            move |result| {
                if let Ok(file) = result {
                    if let Some(path) = file.path() {
                        window.do_export_myexpenses(&path);
                    }
                }
            }
        ));
    }

    fn do_export_myexpenses(&self, zip_path: &Path) {
        let schema_bytes = gio::resources_lookup_data(
            "/io/github/nico359/expenses/myexpenses_schema.sql",
            gio::ResourceLookupFlags::NONE,
        );

        let schema_sql = match schema_bytes {
            Ok(bytes) => String::from_utf8_lossy(&bytes).to_string(),
            Err(e) => {
                self.show_error("Export Failed", &format!("Could not load schema: {}", e));
                return;
            }
        };

        let tmp_dir = std::env::temp_dir();
        let backup_path = tmp_dir.join("expenses-export-BACKUP");

        // Remove old temp file if exists
        std::fs::remove_file(&backup_path).ok();

        match self.create_myexpenses_backup(&backup_path, &schema_sql) {
            Ok(()) => {
                match self.zip_backup(&backup_path, zip_path) {
                    Ok(()) => {
                        self.show_info("Export Complete", "MyExpenses backup created successfully.");
                    }
                    Err(e) => {
                        self.show_error("Export Failed", &format!("Could not create ZIP: {}", e));
                    }
                }
            }
            Err(e) => {
                self.show_error("Export Failed", &format!("Could not create backup: {}", e));
            }
        }

        std::fs::remove_file(&backup_path).ok();
    }

    fn create_myexpenses_backup(&self, backup_path: &Path, schema_sql: &str) -> Result<(), String> {
        let conn = Connection::open(backup_path).map_err(|e| e.to_string())?;
        conn.execute_batch(schema_sql).map_err(|e| e.to_string())?;

        let accounts = self.db().get_accounts().map_err(|e| e.to_string())?;
        let mut account_id_map = std::collections::HashMap::new();

        for (i, account) in accounts.iter().enumerate() {
            let me_id = (i + 1) as i64;
            account_id_map.insert(account.id, me_id);

            // Get opening balance for this account
            let opening: Option<f64> = self.db().connection().query_row(
                "SELECT amount FROM expenses WHERE account_id = ?1 AND is_opening_balance = 1",
                params![account.id],
                |row| row.get(0),
            ).ok();
            let opening_cents = opening.map(|a| (a * 100.0).round() as i64).unwrap_or(0);

            let uuid = uuid::Uuid::new_v4().to_string();
            conn.execute(
                "INSERT INTO accounts (label, opening_balance, description, currency, type, color, uuid) \
                 VALUES (?1, ?2, '', 'EUR', 1, -3355444, ?3)",
                params![account.name, opening_cents, uuid],
            ).map_err(|e| e.to_string())?;
        }

        // Export payees
        let all_expenses = self.db().get_all_expenses_for_export().map_err(|e| e.to_string())?;
        let mut payee_id_map = std::collections::HashMap::new();

        for expense in &all_expenses {
            if !expense.payee.is_empty() && !payee_id_map.contains_key(&expense.payee) {
                conn.execute(
                    "INSERT INTO payee (name, name_normalized) VALUES (?1, ?2)",
                    params![expense.payee, expense.payee.to_lowercase()],
                ).map_err(|e| e.to_string())?;
                let pid = conn.last_insert_rowid();
                payee_id_map.insert(expense.payee.clone(), pid);
            }
        }

        // Export transactions (skip opening balances)
        for expense in &all_expenses {
            if expense.is_opening_balance { continue; }

            let me_account_id = match account_id_map.get(&expense.account_id) {
                Some(id) => *id,
                None => continue,
            };

            let amount_cents = if expense.is_income {
                (expense.amount * 100.0).round() as i64
            } else {
                -(expense.amount * 100.0).round() as i64
            };

            let timestamp = Self::date_str_to_timestamp(&expense.date);
            let payee_id = payee_id_map.get(&expense.payee).copied();
            let uuid = uuid::Uuid::new_v4().to_string();

            conn.execute(
                "INSERT INTO transactions \
                 (comment, date, value_date, amount, account_id, payee_id, cr_status, number, uuid) \
                 VALUES (?1, ?2, ?2, ?3, ?4, ?5, 'UNRECONCILED', '', ?6)",
                params![expense.note, timestamp, amount_cents, me_account_id, payee_id, uuid],
            ).map_err(|e| e.to_string())?;
        }

        Ok(())
    }

    fn zip_backup(
        &self,
        backup_path: &Path,
        zip_path: &Path,
    ) -> Result<(), String> {
        let file = std::fs::File::create(zip_path).map_err(|e| e.to_string())?;
        let mut zip = zip::ZipWriter::new(file);

        let options = zip::write::SimpleFileOptions::default()
            .compression_method(zip::CompressionMethod::Deflated);

        zip.start_file("BACKUP", options).map_err(|e| e.to_string())?;
        let backup_data = std::fs::read(backup_path).map_err(|e| e.to_string())?;
        zip.write_all(&backup_data).map_err(|e| e.to_string())?;
        zip.finish().map_err(|e| e.to_string())?;
        Ok(())
    }

    fn date_str_to_timestamp(date_str: &str) -> i64 {
        // Try "YYYY-MM-DD HH:MM" first, then "YYYY-MM-DD"
        if let Ok(dt) = chrono::NaiveDateTime::parse_from_str(date_str, "%Y-%m-%d %H:%M") {
            dt.and_local_timezone(chrono::Local).unwrap().timestamp()
        } else if let Ok(d) = chrono::NaiveDate::parse_from_str(date_str, "%Y-%m-%d") {
            d.and_hms_opt(0, 0, 0).unwrap()
                .and_local_timezone(chrono::Local).unwrap().timestamp()
        } else {
            chrono::Local::now().timestamp()
        }
    }

    // --- Database Export ---

    pub(crate) fn on_export_db(&self) {
        let timestamp = chrono::Local::now().format("%Y-%m-%d_%H-%M-%S");
        let dialog = gtk::FileDialog::builder()
            .title("Export Database")
            .initial_name(format!("expenses-backup-{}.db", timestamp))
            .build();

        dialog.save(Some(self), None::<&gio::Cancellable>, glib::clone!(
            #[weak(rename_to = window)]
            self,
            move |result| {
                if let Ok(file) = result {
                    if let Some(path) = file.path() {
                        window.do_export_db(&path);
                    }
                }
            }
        ));
    }

    fn do_export_db(&self, dest_path: &Path) {
        match self.db().connection().backup(
            rusqlite::DatabaseName::Main,
            dest_path,
            None,
        ) {
            Ok(()) => {
                // Convert WAL to DELETE journal mode for portability
                if let Ok(dst) = Connection::open(dest_path) {
                    dst.execute_batch("PRAGMA journal_mode=DELETE;").ok();
                }
                self.show_info("Export Complete", "Database exported successfully.");
            }
            Err(e) => {
                self.show_error("Export Failed", &format!("{}", e));
            }
        }
    }

    // --- Database Import ---

    pub(crate) fn on_import_db(&self) {
        let dialog = gtk::FileDialog::builder()
            .title("Import Database")
            .build();

        let filter = gtk::FileFilter::new();
        filter.add_pattern("*.db");
        filter.set_name(Some("SQLite Database"));
        let filters = gio::ListStore::new::<gtk::FileFilter>();
        filters.append(&filter);
        dialog.set_filters(Some(&filters));

        dialog.open(Some(self), None::<&gio::Cancellable>, glib::clone!(
            #[weak(rename_to = window)]
            self,
            move |result| {
                if let Ok(file) = result {
                    if let Some(path) = file.path() {
                        window.confirm_import_db(&path);
                    }
                }
            }
        ));
    }

    fn confirm_import_db(&self, path: &Path) {
        let dialog = adw::AlertDialog::builder()
            .heading("Import Database?")
            .body("This will replace ALL current data with the imported database. This cannot be undone.")
            .build();
        dialog.add_response("cancel", "Cancel");
        dialog.add_response("import", "Import");
        dialog.set_response_appearance("import", adw::ResponseAppearance::Destructive);

        let path = path.to_path_buf();
        dialog.choose(self, None::<&gio::Cancellable>, glib::clone!(
            #[weak(rename_to = window)]
            self,
            move |response| {
                if response == "import" {
                    window.do_import_db(&path);
                }
            }
        ));
    }

    fn do_import_db(&self, path: &Path) {
        match Connection::open_with_flags(path, rusqlite::OpenFlags::SQLITE_OPEN_READ_ONLY) {
            Ok(src) => {
                // Backup source database to the app's database path
                let db_path = Database::db_path();
                match src.backup(rusqlite::DatabaseName::Main, &db_path, None) {
                    Ok(()) => {
                        // Re-initialize database connection
                        self.init_db();
                        self.setup_accounts();
                        self.refresh_expenses();
                        self.show_info("Import Complete", "Database imported successfully.");
                    }
                    Err(e) => {
                        self.show_error("Import Failed", &format!("{}", e));
                    }
                }
            }
            Err(e) => {
                self.show_error("Import Failed", &format!("Could not open file: {}", e));
            }
        }
    }

    // --- Helper dialogs ---

    fn show_info(&self, heading: &str, body: &str) {
        let dialog = adw::AlertDialog::builder()
            .heading(heading)
            .body(body)
            .build();
        dialog.add_response("ok", "OK");
        dialog.present(Some(self));
    }

    fn show_error(&self, heading: &str, body: &str) {
        let dialog = adw::AlertDialog::builder()
            .heading(heading)
            .body(body)
            .build();
        dialog.add_response("ok", "OK");
        dialog.present(Some(self));
    }
}

